### PR TITLE
LF-3583: Create select revenue type container and add to routes

### PR DESCRIPTION
**Pull request by SayakaOno**
**Description**

NOTE:
- selected revenue type will not be saved in the DB yet. (it will be stored in the Redux store)
- clicking any type will be navigated to the form for crop sale.

To see types, please use Joyce's `interim-revenue-type-API` [branch](https://github.com/LiteFarmOrg/LiteFarm/tree/interim-revenue-type-API). (Thank you Joyce!)

Query to add some custom types. (Please replace `<<farm_id>>` with farm_id of your farm)
```
INSERT INTO "public"."revenue_type"("revenue_name", "farm_id", "created_at", "updated_at", "revenue_translation_key")
VALUES
  ('Service', '<<farm_id>>', '1999-12-31 16:00:00-08', '1999-12-31 16:00:00-08', 'SERVICE'),
  ('CSA Box', '<<farm_id>>', '1999-12-31 16:00:00-08', '1999-12-31 16:00:00-08', 'CSA_BOX'),
  ('Jam', '<<farm_id>>', '1999-12-31 16:00:00-08', '1999-12-31 16:00:00-08', 'JAM'),
  ('Drink', '<<farm_id>>', '1999-12-31 16:00:00-08', '1999-12-31 16:00:00-08', 'DRINK');
  ```

- create `RevenueTypes` container
- create `useSortedRevenueTypes` hook
- update `onClick` of "Add New Sale" button to navigate to the revenue types page
- update [stories](http://localhost:6006/?path=/story/page-finance-purefinancetypeselection--revenue)

Jira link: https://lite-farm.atlassian.net/browse/LF-3583

**Type of change**

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

- [ ] Passes test case
- [x] UI components visually reviewed on desktop view
- [x] UI components visually reviewed on mobile view
- [ ] Other (please explain)

**Checklist:**

- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] The precommit and linting ran successfully
- [x] I have added or updated language tags for text that's part of the UI
- [x] I have added "MISSING" for all new language tags to languages I don't speak
- [x] I have added [the GNU General Public License](https://lite-farm.atlassian.net/l/cp/BT0Dd7WW) to all new files

### CODE REVIEW

File: packages/webapp/src/containers/Finances/AddSale/RevenueTypes/useSortedRevenueTypes.jsx [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+  useEffect(() => {
+    dispatch(getRevenueTypes());
+  }, []);
+
+  useEffect(() => {
+    const defaultTypes = [];
+    const customTypes = [];
+    revenueTypes?.forEach((type) => {
+      if (!type.deleted) {
```
Comment (kathyavini) [inline packages/webapp/src/containers/Finances/AddSale/RevenueTypes/useSortedRevenueTypes.jsx:37]: I don't mind an extra check at all but that particular selector does already filter for deleted, so it's not... possible, right? That `type.deleted` could be true?

And ooops in checking this I just realized I have a useless destructuring of `deleted` in that selector that I didn't notice before 😬

**Review by kathyavini — APPROVED** [review]
I really love the hook!!

Very clean and works great!

File: packages/webapp/src/containers/Finances/AddSale/RevenueTypes/useSortedRevenueTypes.jsx (hunk 1 above)
Comment (SayakaOno) [reply packages/webapp/src/containers/Finances/AddSale/RevenueTypes/useSortedRevenueTypes.jsx:37]: Thank you Joyce!

No, you are right, I used a selector that selects all types including deleted ones for expense type, and I forgot to remove the check from the code I copied... I will remove it!
I spotted the unused `deleted` but forgot to remove. Shall I remove it? If you made a commit already, I will leave it!

Comment (kathyavini) [reply packages/webapp/src/containers/Finances/AddSale/RevenueTypes/useSortedRevenueTypes.jsx:37]: Would you remove it for me?? 😍 No I didn't make a commit!

Comment (SayakaOno) [reply packages/webapp/src/containers/Finances/AddSale/RevenueTypes/useSortedRevenueTypes.jsx]: Done 😄

**Review by kathyavini — APPROVED** [review]
Thank you! 🙏

File: packages/webapp/src/containers/Finances/AddSale/RevenueTypes/useSortedRevenueTypes.jsx [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+ *  GNU General Public License for more details, see <https://www.gnu.org/licenses/>.
+ */
+import { useEffect, useState } from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+import PropTypes from 'prop-types';
+import { useTranslation } from 'react-i18next';
+import { getRevenueTypes } from '../../saga';
+import { revenueTypesSelector } from '../../../revenueTypeSlice';
+
+export default function useSortedRevenueTypes() {
```
Comment (Duncan-Brain) [inline packages/webapp/src/containers/Finances/AddSale/RevenueTypes/useSortedRevenueTypes.jsx:22]: @SayakaOno I will steal your question haha sorry .. Is there a reason you chose not to make this custom hook reusable?

**Comment by Duncan-Brain** [comment]
I can approve I just had one question I thought I should check first.

1. Since we cannot select multiple revenue types, is this view part of the form like expenses? And then is the form persist necessary here? It would still make sense as a persisted path for the next view -- crop sale. Or maybe you are adding it now for the case where multiple revenue types can be selected in the future?

File: packages/webapp/src/containers/Finances/AddSale/RevenueTypes/useSortedRevenueTypes.jsx (hunk 2 above)
Comment (SayakaOno) [reply packages/webapp/src/containers/Finances/AddSale/RevenueTypes/useSortedRevenueTypes.jsx:22]: How do you want to make it reusable? By passing getTypes action, selector and those `revenue:${typeA.revenue_translation_key}`? I prefer to make another hook in this case.

Sorry "reusable" was very wrong word choice, but I was expecting something simpler that could be used only for these two components like this:
```
const validateUniqueTypeName = (value) => {
  const expenseNameExists = expenseTypes.some((type) => {
    return type.expense_name === value;
  });

  if (expenseNameExists) {
    return t('EXPENSE.ADD_EXPENSE.DUPLICATE_NAME');
  }
  return true;
};
```
⬇️ 
```
export const validateUniqueTypeName = (expenseTypes, value) => {
  const expenseNameExists = expenseTypes.some((type) => {
    return type.expense_name === value;
  });

  if (expenseNameExists) {
    return t('EXPENSE.ADD_EXPENSE.DUPLICATE_NAME');
  }
  return true;
};
```

**Comment by SayakaOno** [comment]
What is another way to pass the selected revenue type to the next view?
`history.push(..., {state: {revenue_type_id}})` or something?

File: packages/webapp/src/containers/Finances/AddSale/RevenueTypes/useSortedRevenueTypes.jsx (hunk 2 above)
Comment (Duncan-Brain) [reply packages/webapp/src/containers/Finances/AddSale/RevenueTypes/useSortedRevenueTypes.jsx:22]: Okay sounds good haha! I figured maybe the same logic was used for expense types then maybe task types etc. But I thought I should ask!

Comment (SayakaOno) [reply packages/webapp/src/containers/Finances/AddSale/RevenueTypes/useSortedRevenueTypes.jsx:22]: Ah that's true! I think it's good to have a reusable sorting function at least!

**Comment by Duncan-Brain** [comment]
Maybe that might work.... I do now see that this is how Task Type selection works. But we are also aware it does cause white screen if you navigate there without being on a persisted path:
Screenshot 2023-09-08 at 3 43 20 PM

For single page forms I don't mind the pattern since all the data is available in url:
`history.push('/add_revenue/:revenue_type_id')`

But yah just checking -- happy to approve as is!

**Review by Duncan-Brain — APPROVED** [review]

**Comment by Duncan-Brain** [comment]
Also sorry @SayakaOno and @kathyavini when running the insert query above to test I got an index error ( I think because we started at 1 instead of 0? in the migration to create the table). 

Did either of you two receive the same error? It worked the second time.

**Comment by kathyavini** [comment]
@Duncan-Brain I have indeed seen an index error on that table when I added a record manually! I moved to using the add saga and didn't have an issue; I wonder if the success was just an artifact of having tried and failed once before then...

**Comment by SayakaOno** [comment]
@Duncan-Brain I'm a supporter of `/add_revenue`, but we can discuss it in the future :)
I don't think I got an index error...

**Pull request merged**